velox/exec/tests/SparkShuffleTest.cpp

```diff
@@ -0,0 +1,611 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "velox/exec/Exchange.h"
+#include "velox/exec/HashPartitionFunction.h"
+#include "velox/exec/tests/PartitionAndSerialize.h"
+#include "velox/exec/tests/ShuffleWrite.h"
+#include "velox/exec/tests/utils/OperatorTestBase.h"
+#include "velox/exec/tests/utils/PlanBuilder.h"
+#include "velox/expression/VectorFunction.h"
+#include "velox/serializers/UnsafeRowSerde.h"
+#include "velox/serializers/UnsafeRowSerializer.h"
+
+using namespace facebook::velox;
+
+namespace facebook::velox::exec::test {
+
+namespace {
+class PartitionFunction : public exec::VectorFunction {
+ public:
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& outputType,
+      EvalCtx* context,
+      VectorPtr* result) const override {
+    // TODO Verify that partition count is constant.
+    auto numPartitions =
+        args[0]->as<SimpleVector<int32_t>>()->valueAt(rows.begin());
+
+    auto rowType = makeRowType(args);
+
+    auto argsCopy = args;
+    auto input = std::make_shared<RowVector>(
+        context->pool(), rowType, nullptr, rows.size(), std::move(argsCopy));
```

Comment:
Not sure if we need move here. Can we just pass ```args``` so it creates a copy?

Reply:
It is not necessary, but saves a vector copy.

Reply:
I don't believe we can pass 'args' because it is a reference, but the RowVector constructor expect a value.